tests/specs/mcd/verification.k

```diff
@@ -109,4 +119,53 @@ module VERIFICATION
+
+    // Proof:
+    // case sgn(X) ==Int  1
+    //      X s<Word 0
+    //   => X <Word 0           // sgn(X) ==Int 1 andBool sgn(0) ==Int 1
+    //   => bool2Word(false)    // #rangeUInt(256, X)
+    // case sgn(X) ==Int -1
+    //      X s<Word 0
+    //   => bool2Word(true)     // sgn(X) ==Int -1 andBool sgn(0) ==Int 1
+    rule X s<Word 0 => bool2Word(sgn(X) ==Int -1) requires #rangeUInt(256, X)
+
+    rule sgn(chop(X)) ==Int -1 => X <Int 0 requires #rangeSInt(256, X)
+
+    // ### vat-mului-pass
+
+    rule chop(X *Int chop(Y)) => chop(X *Int Y)
+
+    // Proof #rangUInt(256, X) andBool #rangeSInt(256, Y) andBool #rangeSInt(256, X *Int Y) andBool Y =/=Int 0
+    //      chop(X *Int Y) /sWord chop(Y)
+    //   => #sgnInterp(sgn(chop(X *Int Y)) *Int sgn(chop(Y)), abs(chop(X *Int Y)) /Word abs(chop(Y)))
+    // Focus: sgn(chop(X *Int Y)) *Int sgn(chop(Y))
+    //      case 0 <=Int chop(Y)
+    //           sgn(chop(X *Int Y)) *Int sgn(chop(Y))
+    //        => sgn(chop(X *Int Y)) *Int sgn(Y)                     // chop(Y) ==Int Y
+    //        => sgn(X *Int Y) *Int sgn(Y)                           // 0 <=Int X andBool 0 <=Int Y
+    //        => 1 *Int 1                                            // 0 <=Int X *Int Y andBool 0 <=Int Y
+    //        => 1
+    //      case Y <Int 0
+    //           sgn(chop(X *Int Y)) *Int sgn(chop(Y))
+    //        => sgn(X *Int Y +Int pow256) *Int sgn(Y +Int pow256)   // chop(Y) ==Int Y +Int pow256
+    //        => -1 *Int -1                                          // X *Int Y <Int 0 andBool Y <Int 0
+    //        => 1
+    // Focus: abs(chop(X *Int Y)) /Word abs(chop(Y))
+    //      case 0 <=Int Y
+    //           abs(chop(X *Int Y)) /Word abs(chop(Y))
+    //        => abs(chop(X *Int Y)) /Int  abs(chop(Y))
+    //        => abs(X *Int Y) /Int abs(Y)                                                   // 0 <=Int X andBool 0 <=Int Y andBool #rangeSigne andBool chop(Y) ==Int Y
+    //        => X *Int Y /Int Y                                                             // #rangeSInt(256, X *Int Y) andBool 0 <=Int X *Int Y andBool #rangeSInt(256, Y) andBool 0 <=Int Y
+    //        => X
+    //      case Y <Int 0
+    //           abs(chop(X *Int Y)) /Int abs(chop(Y))
+    //        => abs(X *Int Y +Int pow256) /Int abs(Y +Int pow256)                           // #rangeSInt(256, X *Int Y) andBool chop(Y) ==Int Y +Int pow256
+    //        => (0 -Word (X *Int Y +Int pow256)) /Int (0 -Word (Y +Int pow256))             // #rangeSInt(256, X *Int Y) andBool X *Int Y <Int 0 andBool #rangeSInt(256, Y) andBool Y <Int 0
+    //        => (pow256 -Int (X *Int Y +Int pow256)) /Int (pow256 -Int (Y +Int pow256))     // 0 <Int (X *Int Y +Int pow256) andBool 0 <Int (Y +Int pow256)
+    //        => (X *Int Y) /Int Y
+    //        => X
+    rule chop(X *Int Y) /sWord chop(Y) => X requires #rangeUInt(256, X) andBool #rangeSInt(256, Y) andBool #rangeSInt(256, X *Int Y) andBool Y =/=Int 0
+
```

Comment:
Have you tried to prove this using z3?

Reply:
A partial proof with a small model (for 6-bit integers): https://rise4fun.com/Z3/Eslc8

Provable by z3 4.6.0 or 4.7.1, but not 4.8.x.
(The z3 in the web interface failed to prove.)

Reply:
Although I didn't read the hand-written proof, since z3 proved the lemma with a smaller bitwidth, and the lemma doesn't seem dependent on the particular bitwidth, I think the lemma is fine.

Reply:
Hmmm, I can't get this Z3 proof to work on any version of Z3 I've tried, even though it looks OK.

Reply:
Have you tried 4.6.0 or 4.7.1?